geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java

```diff
@@ -2302,9 +2294,9 @@ void checkQueueSizeConstraint() throws InterruptedException {
      * @return int current Put permits
      */
     private int reconcilePutPermits() {
-      putPermits += takeSidePutPermits;
-      takeSidePutPermits = 0;
-      return putPermits;
+      putPermits.addAndGet(takeSidePutPermits.get());
```

Comment:
I think you have a concurrency bug between takeSidePutPermits.get() and the next line takeSidePutPermits.set(0).
If some other thread incs takeSidePutPermits between these two statements then the inc is lost.
I think you want to rewrite this code like so:

int takes = takeSidePutPermits.getAndSet(0);
return putPermits.addAndGet(takes);

Reply:
Yup.